[SymmMem] Allow selection of allocation backend

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):

Today the only way to choose allocation backend is via env `TORCH_SYMMMEM=...`.
This is a bit hard to set in CI on test file basis. (The env has to be set before program is loaded).

This PR added a programmatic way -- a `set_backend` API.

Implementation:
Since this API is slightly more dynamic than static registration, at static time each backend registers its availability rather than filling itself as **the** allocator directly. Later when `set_backend` is called, the allocator would actually fill in the device-to-allocation `map_`.

Though added, `set_backend` is **not** a necessary API for user to call -- one backend is still registered as the default at static time. 

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

kwen2501 committed:
Update

[ghstack-poisoned]

fduwjj reviewed (commented):
what if user call this set_backend() more than once for CUDA, for example symm_mem.set_backend("NVSHMEM") then symm_mem.set_backend("NCCL") and then symm_mem.set_backend("NVSHMEM"). Is this defined?

kwen2501 commented:
@fduwjj The API's own behavior is still defined per the API's doc string. 
For consequence further down in the application, like whether one would be able to allocate memory from two different allocators, I don't know, can't define it atm.

kwen2501 committed:
Update on "[SymmMem] Allow selection of allocation backend"

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):

Today the only way to choose allocation backend is via env `TORCH_SYMMMEM=...`.
This is a bit hard to set in CI on test file basis. (The env has to be set before program is loaded).

This PR added a programmatic way -- a `set_backend` API.

Implementation:
Since this API is slightly more dynamic than static registration, at static time each backend registers its availability rather than filling itself as **the** allocator directly. Later when `set_backend` is called, the allocator would actually fill in the device-to-allocation `map_`.

Though added, `set_backend` is **not** a necessary API for user to call -- one backend is still registered as the default at static time. 

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 commented:
Thanks @fduwjj @ngimel .
I added an `in_use_` flag to forbid change of backend after allocation.
Can you please take a look when you have another chance? Thanks.

ngimel commented on torch/distributed/_symmetric_memory/__init__.py:
```diff
@@ -1720,4 +1720,18 @@ def is_nvshmem_available() -> bool:
 
 
-__all__ = ["empty", "rendezvous", "is_nvshmem_available"]
+def set_backend(name: Literal["NVSHMEM", "CUDA", "NCCL"]) -> None:
```
can you also add `get_backend` while you are at it?

ngimel reviewed (commented):
This lgtm, btw what happens if there's a build without NVSHMEM but user requests nvshmem backend? I take it CUDA and NCCL are universally available (though I have doubts about nccl in fbcode)

kwen2501 commented:
@ngimel I added an is_nvshmem_available API previously:

Add runtime detection of NVSHMEM (PR #156291)

kwen2501 committed:
Update on "[SymmMem] Allow selection of allocation backend"

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):

Today the only way to choose allocation backend is via env `TORCH_SYMMMEM=...`.
This is a bit hard to set in CI on test file basis. (The env has to be set before program is loaded).

This PR added a programmatic way -- a `set_backend` API.

Implementation:
Since this API is slightly more dynamic than static registration, at static time each backend registers its availability rather than filling itself as **the** allocator directly. Later when `set_backend` is called, the allocator would actually fill in the device-to-allocation `map_`.

Though added, `set_backend` is **not** a necessary API for user to call -- one backend is still registered as the default at static time. 

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge

fduwjj commented on torch/csrc/distributed/c10d/symm_mem/SymmetricMemory.cpp:
```diff
@@ -22,13 +22,47 @@ class AllocatorMap {
         it != map_.end(),
         "SymmetricMemory does not support device type ",
         device_type);
+    in_use_ = true;
```
ideally this flag is per device?

kwen2501 commented:
@pytorchbot merge -f

kwen2501 commented:
@pytorchbot merge -f "2 Unrelated Failures"
